p2p/src/codec.rs

```diff
@@ -0,0 +1,268 @@
+use crate::cybershake::PublicKey;
+use crate::peer::PeerAddr;
+use crate::{PeerID, PeerMessage};
+use bytes::{Buf, BufMut, BytesMut};
+use curve25519_dalek::ristretto::CompressedRistretto;
+use std::io;
+use std::net::{Ipv4Addr, Ipv6Addr, SocketAddr, SocketAddrV4, SocketAddrV6};
+use tokio_util::codec::{Decoder, Encoder};
+
+pub struct MessageEncoder;
+
+impl Encoder<PeerMessage> for MessageEncoder {
+    type Error = io::Error;
+
+    fn encode(&mut self, item: PeerMessage, dst: &mut BytesMut) -> Result<(), Self::Error> {
+        match item {
+            PeerMessage::Hello(u) => {
+                dst.put_u8(0); // Message type
+                const HELLO_MESSAGE_LEN: u64 = 2;
+                dst.put_u64(HELLO_MESSAGE_LEN);
+                dst.put_u16(u);
+            }
+            PeerMessage::Peers(p) => {
+                dst.put_u8(1); // Message type
+                let body = p.into_iter().fold(BytesMut::new(), |mut bytes, peer| {
+                    encode_peer_addr(peer, &mut bytes);
+                    bytes
+                });
+                dst.put_u64(body.len() as u64);
+                dst.put(body.as_ref());
+            }
+            PeerMessage::Data(data) => {
+                dst.put_u8(2); // Message type
+                dst.put_u64(data.len() as u64);
+                dst.put(data.as_slice());
+            }
+        }
+        Ok(())
+    }
+}
+
+fn encode_peer_addr(peer: PeerAddr, buf: &mut BytesMut) {
```

Comment:
i'd put this function close to decoding function, so it's easier to review their consistency.

Reply:
Done